extkeys/mnemonic.go

```diff
@@ -216,17 +232,61 @@ func (m *Mnemonic) ValidMnemonic(mnemonic string, language Language) bool {
 
 	// The number of words should be 12, 15, 18, 21 or 24
 	if numOfWords%3 != 0 || numOfWords < 12 || numOfWords > 24 {
-		return false
+		return errors.New("mnemonic contains an invalid number of words")
 	}
 
+	// Create reverse lookup map for dictionary
+	wordMap := map[string]int{}
+	for i, v := range wordList {
+		wordMap[v] = i
+	}
+
+	b := big.NewInt(0)
+
 	// Check if all words belong in the wordlist
 	for i := 0; i < numOfWords; i++ {
-		if !contains(wordList, words[i]) {
-			return false
+
+		wordIndex, ok := wordMap[words[i]]
+		if !ok {
+			return errors.New("not all words are found in dictionary")
 		}
+		var wordBytes [2]byte
+		binary.BigEndian.PutUint16(wordBytes[:], uint16(wordIndex))
+		b = b.Mul(b, rightShift11BitsDivider)
+		b = b.Or(b, big.NewInt(0).SetBytes(wordBytes[:]))
+	}
+
+	checksum := big.NewInt(0)
+	checksumMask := wordLengthChecksumMasksMapping[len(words)]
```

Comment:
This can return `nil` and it's not checked in the line below. This can result in panic. We either make sure this case is tested or we do a check.

```suggestion
	checksumMask := wordLengthChecksumMasksMapping[numOfWords]
```

Reply:
We check earlier in the code to verify that the number of words equals one of the acceptable values and return an error if not.
```
if numOfWords%3 != 0 || numOfWords < 12 || numOfWords > 24 {
		return errors.New("mnemonic contains an invalid number of words")
	}
```

Is it still possible we get a `nil` back from this line?

Reply:
No, unless one shuffles the code around :)